digiroad2-api-oth/src/main/scala/fi/liikennevirasto/digiroad2/dataimport/CsvDataImporter.scala

```diff
@@ -110,26 +119,73 @@ class TrafficSignCsvImporter extends CsvDataImporterOperations {
   }
 
   private def verifyValueCode(parameterName: String, parameterValue: String): ParsedAssetRow = {
-    if(parameterValue.forall(_.isDigit) && TRTrafficSignType.apply(parameterValue.toInt).source.contains("CSVimport")){
+    if(parameterValue.forall(_.isDigit) && TrafficSignType.applyTRValue(parameterValue.toInt).source.contains("CSVimport")){
       (Nil, List(AssetProperty(columnName = codeValueFieldMappings(parameterName), value = parameterValue.toInt)))
     }else{
       (List(parameterName), Nil)
     }
   }
 
-  def tryToInt(propertyValue: String ) = {
+  def getRightRoadLinkUsingBearing(assetBearing: Option[Int], assetCoordinates: Point): (Seq[VVHRoadlink], Seq[RoadLink]) = {
+    val toleranceInDegrees = 25
+    val roadLinks = roadLinkService.getClosestRoadlinkForCarTrafficFromVVH(userProvider.getCurrentUser(), assetCoordinates)
+
+    assetBearing match {
+      case Some(aBearing) if roadLinks.nonEmpty =>
+        val enrichedRoadLinks = roadLinkService.enrichRoadLinksFromVVH(roadLinks)
+
+        val filteredEnrichedRoadLinks =
+          enrichedRoadLinks.filter { roadLink =>
+            val aTrafficDirection = TrafficDirection.apply(trafficSignService.getTrafficSignValidityDirection(assetCoordinates, roadLink.geometry))
+
+            val mValue = GeometryUtils.calculateLinearReferenceFromPoint(assetCoordinates, roadLink.geometry)
+            val roadLinkBearing = GeometryUtils.calculateBearing(roadLink.geometry, Some(mValue), Some(roadLink.length) )
+
+            if (roadLink.trafficDirection == TrafficDirection.BothDirections) {
+              val reverseRoadLinkBearing =
+                if (roadLinkBearing - 180 < 0) {
+                  roadLinkBearing + 180
+                } else {
+                  roadLinkBearing - 180
+                }
+
+              Math.abs(aBearing - roadLinkBearing) <= toleranceInDegrees ||
+                Math.abs(aBearing - reverseRoadLinkBearing) <= toleranceInDegrees
+            } else {
+              Math.abs(aBearing - roadLinkBearing) <= toleranceInDegrees &&
+                (roadLink.trafficDirection == aTrafficDirection || aTrafficDirection == TrafficDirection.BothDirections)
+            }
+          }
+
+        if (filteredEnrichedRoadLinks.nonEmpty) {
+          (roadLinks.filter { rl => filteredEnrichedRoadLinks.exists(_.linkId == rl.linkId) }, filteredEnrichedRoadLinks)
+        } else
+          (roadLinks, Seq())
+
+      case _ =>
+        (roadLinks, Seq())
+    }
+  }
+
```

Comment:
This method should only return a Seq[VVHRoadLink], and probably something like a boolean, because the sequence can return empty and have to possible reasons why: one is because there's no authorization on municipality or no nearby links and the other is because the bearing differs to much from the nearest road link bearing. The boolean could allows us to check the reason on why it's empty so we could proceed or not with the rest of the process.

Reply:
You need to see if it's always needed the enrichedRoadLinks (RoadLinks) on the future of the process.
But I think if you put a flag like "toBeFloating" and validate on the head and if it's true only insert them as a floating and I think we don't have problems anymore.